src/coreclr/src/tools/Common/Compiler/CompilerTypeSystemContext.cs

```diff
@@ -330,6 +331,68 @@ private PdbSymbolReader OpenAssociatedSymbolFile(string peFilePath, PEReader peR
 
             return reader;
         }
+
+        private object _normalizedLock = new object();
+        private HashSet<TypeDesc> _nonNormalizedTypes = new HashSet<TypeDesc>();
+        private Dictionary<TypeDesc, TypeFlags> _normalizedTypeCategory = new Dictionary<TypeDesc, TypeFlags>();
+
+        public TypeFlags NormalizedCategoryFor4ByteStructOnX86(TypeDesc type)
+        {
+            lock(_normalizedLock)
+            {
+                if (_nonNormalizedTypes.Contains(type))
+                    return type.Category;
+                
+                if (_normalizedTypeCategory.TryGetValue(type, out TypeFlags category))
+                    return category;
+
+                if (Target.Architecture != TargetArchitecture.X86)
+                {
+                    throw new NotSupportedException();
+                }
+
+                if ((type.Category != TypeFlags.ValueType) || (type.GetElementSize().AsInt != 4))
+                {
+                    _nonNormalizedTypes.Add(type);
+                    return type.Category;
+                }
+
+                TypeDesc typeOfEmbeddedField = null;
+                foreach (var field in type.GetFields())
+                {
+                    if (field.IsStatic)
+                        continue;
+                    if (typeOfEmbeddedField != null)
+                    {
+                        // Type has more than one instance field
+                        _nonNormalizedTypes.Add(type);
+                        return type.Category;
+                    }
+
+                    typeOfEmbeddedField = field.FieldType;
+                }
+
+                if ((typeOfEmbeddedField.IsValueType) || (typeOfEmbeddedField.IsPointer))
```

Comment:
Would we get here with a null `typeOfEmbeddedField` for `[StructLayout(Explicit, Size=4)]struct Foo { }`?

Reply:
Sigh, yes, that's a bug.